rasterio/_vsiopener.pyx

```diff
@@ -0,0 +1,160 @@
+# the plugin to determine what "files" exist on "disk".
+# Currently the only way to "create" a file in the filesystem is to add
+# an entry to this dictionary. GDAL will then Open the path later.
+cdef _OPENER_REGISTRY = {}
+cdef _OPEN_FILE_OBJS = set()
+
+
+cdef int install_pyopener_plugin(VSIFilesystemPluginCallbacksStruct *callbacks_struct):
+    """Install handlers for python file openers if it isn't already installed."""
+    cdef char **registered_prefixes = VSIGetFileSystemsPrefixes()
+    cdef int prefix_index = CSLFindString(registered_prefixes, PREFIX_BYTES)
+    CSLDestroy(registered_prefixes)
+
+    if prefix_index < 0:
+        log.debug("Installing Python opener handler plugin...")
+        callbacks_struct = VSIAllocFilesystemPluginCallbacksStruct()
+        callbacks_struct.open = <VSIFilesystemPluginOpenCallback>pyopener_open
+        callbacks_struct.tell = <VSIFilesystemPluginTellCallback>pyopener_tell
+        callbacks_struct.seek = <VSIFilesystemPluginSeekCallback>pyopener_seek
+        callbacks_struct.read = <VSIFilesystemPluginReadCallback>pyopener_read
+        callbacks_struct.close = <VSIFilesystemPluginCloseCallback>pyopener_close
+        callbacks_struct.pUserData = <void*>_OPENER_REGISTRY
+        retval = VSIInstallPluginHandler(PREFIX_BYTES, callbacks_struct)
+        VSIFreeFilesystemPluginCallbacksStruct(callbacks_struct)
+        return retval
+    else:
+        return 0
+
+
+cdef void uninstall_pyopener_plugin(VSIFilesystemPluginCallbacksStruct *callbacks_struct):
+    if callbacks_struct is not NULL:
+        callbacks_struct.pUserData = NULL
+        VSIFreeFilesystemPluginCallbacksStruct(callbacks_struct)
+    callbacks_struct = NULL
+
+
+cdef void* pyopener_open(void *pUserData, const char *pszFilename, const char *pszAccess) with gil:
+    """Access files in the virtual filesystem.
+
+    This function is mandatory in the GDAL Filesystem Plugin API.
+    GDAL may call this function multiple times per filename and each
+    result must be seperately seekable.
+    """
+    if pszAccess != b"r" and pszAccess != b"rb":
+        log.error("Python opener is currently a read-only interface.")
+        return NULL
+
+    if pUserData is NULL:
+        log.error("Python opener registry is not initialized.")
+        return NULL
+
+    cdef dict registry = <object>pUserData
+    filename = pszFilename.decode("utf-8")
+
+    log.debug("Looking up opener: registry=%r, filename=%r", registry, filename)
+    try:
+        file_opener = registry[filename]
+    except KeyError:
+        log.info("Object not found: registry=%r, filename=%r", registry, filename)
+        return NULL
+
+    cdef object file_obj
+
+    try:
+        file_obj = file_opener(filename, mode="rb")
+    except ValueError:
+        file_obj = file_opener(filename, mode="r")
+
+    # If the opener returned an OpenFile, we open it.
+    if not hasattr(file_obj, "f") and hasattr(file_obj, "fs") and hasattr(file_obj, "path"):
```

Comment:
I'm a little confused as to the logic here. Why does fsspec need to be a special case here?

Using `rasterio.open(..., opener=fsspec.open)` seems like it violates the expected interface for opener (which is to return an object with read()/seek()/tell()/close() methods).

Perhaps a wrapper function for fsspec might be better suited as the opener function?
```python
def fsspec_open(fn, mode='rb'):
    return fsspec.open(fn, mode=mode).open()
```

Reply:
@groutr I'm trying to be pragmatic, because fsspec is part of our core users' toolkit. The result of `fsspec.open()` is very nearly a file object, it only requires entering its context to get a "live" file object.

Reply:
For the sake of simplifying the code here, I currently agree with @groutr at the moment and suggest adding documentation for recommended usage patterns when using `opener`. I think with the frequency of using `functools.partial` for custom arguments, having a function wrapper wouldn't be too much of an ask.

Reply:
It may be worthwhile to raise an issue with `fsspec` to see if they could make changes to make their file object more compatible?

Reply:
With a nicely designed opener function for fsspec, we could probably side-step most of the need to use functools.partial.

```python
def fsspec_with(**kwargs):
    def opener(fn):
        return fsspec.open(fn, mode='rb', **kwargs)
    return opener
```
Then you could make:
```python
from functools import partial
with rasterio.open("dask::s3://bucket/key", opener=partial(fsspec.open, s3={"anon": True})) as src:
    print(src.profile)
```
into:
```python
with rasterio.open("dask::s3://bucket/key", opener=fsspec_with(s3={"anon": True}) as src:
    print(src.profile)
```

Reply:
I got a solution coming up that removes the `OpenFile` special case.